src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.cs

```diff
@@ -216,10 +216,43 @@ private static void GenerateWriteableField(ArrayBuilder<CodeAction> result, Sema
                 if (containingMethod.Parameters.Length > 0)
                 {
                     var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-                    var lastParameterIsCancellationToken = Equals(containingMethod.Parameters.Last().Type, compilation.CancellationTokenType());
+                    var cancellationTokenType = compilation.CancellationTokenType();
 
-                    if (lastParameterIsCancellationToken)
-                        parameterIndex--;
+                    for (var i = containingMethod.Parameters.Length - 1; i >= 0; i--)
+                    {
+                        var parameter = containingMethod.Parameters[i];
+
+                        // Keep moving the insertion position for the generated parameter backwards
+                        // until we get to a parameter that does not need to be at the end of the
+                        // parameter list.
+                        if (parameter.HasExplicitDefaultValue)
+                        {
+                            parameterIndex--;
+                        }
+                        else if (parameter.IsParams)
+                        {
+                            parameterIndex--;
+                        }
+                        else if (parameter.RefKind is RefKind.Out)
+                        {
+                            parameterIndex--;
+                        }
+                        else if (Equals(parameter.Type, cancellationTokenType))
+                        {
+                            parameterIndex--;
+                        }
+                        else
+                        {
+                            break;
+                        }
+                    }
+
+                    // If we are in an extension method, then we want to make sure to insert after
+                    // the first parameter.
+                    if (containingMethod.IsExtensionMethod && parameterIndex == 0)
+                    {
+                        parameterIndex++;
+                    }
```

Comment:
Nit:  Similarly:

```suggestion
                    if (containingMethod.IsExtensionMethod && parameterIndex == 0)
                    {
                        parameterIndex = 1;
                    }
```

The ++ makes it seem like we're really trying to update this relative to the last value it had.  But that's not really waht's going on.  We aren't saying "read the last value and add one to it to find the value ot use".  We're realy just saying "if this is an extension method, any added parameters cannot start before index 1.